### PR TITLE
Templated arm64 integration and e2e workflows for main and release-3.5

### PR DESCRIPTION
**Pull request by jmhbnz**
Follow-on from https://github.com/etcd-io/etcd/pull/16151 to setup scheduled workflows for `arm64` nightly integration and e2e suites against the `release-3.5` branch.

Arm64 integration and e2e suites now adopt a shared re-usable workflow template which is slightly customised for `main` or `release-3.5`.

Fixes https://github.com/etcd-io/etcd/issues/15832

### CODE REVIEW

**Comment by jmhbnz** [comment]
Note for reviewers: Workflows have been tested `on: pull_request` and were green for both `main` and `release-3.5`:
- e2e: https://github.com/etcd-io/etcd/actions/runs/5409330146
- integration: https://github.com/etcd-io/etcd/actions/runs/5409330130

File: .github/workflows/tests-arm64-nightly.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    with:
+      etcdBranch: main
+      integrationTestCmd: make test-integration
+      unitTestCmd: GO_TEST_FLAGS='-p=2' make test-unit
+  release-35-arm64:
+    uses: ./.github/workflows/tests-arm64-template.yaml
+    with:
+      etcdBranch: release-3.5
+      integrationTestCmd: PASSES='integration' RACE='false' ./test.sh
+      unitTestCmd: PASSES='unit' CPU='4' ./test.sh -p=2
```
Comment (serathius) [inline .github/workflows/tests-arm64-nightly.yaml:21]: Could we maybe unify the commands to make it clearer?

File: .github/workflows/e2e-arm64-nightly.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+---
```
Comment (serathius) [inline .github/workflows/e2e-arm64-nightly.yaml:1]: With addition of ARM nightly per branch I think we are getting the limit of clarity of `./github/workflows` directory.
Maybe we could consider creating subdirs and grouping workflows by purpose:
* `./github/workflows/periodics`
* `./github/workflows/presubmits`

Leaving the a idea for a followup.

File: .github/workflows/tests-arm64-nightly.yaml (hunk 1 above)
Comment (jmhbnz) [reply .github/workflows/tests-arm64-nightly.yaml:21]: Can you please clarify.  Do you mean backporting the makefile changes so both `release-3.5` and `main` can use same `make` command?  Currently `release-3.5` seems to need `./test.sh`.

File: .github/workflows/tests-arm64-template.yaml [hunk 3]
```diff
@@ -40,19 +56,19 @@ jobs:
           export JUNIT_REPORT_DIR=$(realpath ${TARGET})
           case "${TARGET}" in
             linux-arm64-integration-1-cpu)
-              make gofail-enable
-              GOOS=linux GOARCH=arm64 CPU=1 make test-integration
+              if [ "${{ inputs.gofailMake }}" == "yes" ]; then make gofail-enable; fi
```
Comment (serathius) [inline .github/workflows/tests-arm64-template.yaml:59]: I'm worried that such forced compatibility will make it harder to add/update tests on main branch and will not help at all with preventing a breaking change. 

I think I would just to prefer to keep the definition of  periodic and pull request workflows separate.

Comment (jmhbnz) [reply .github/workflows/tests-arm64-template.yaml:59]: These templates are only used for periodic.  We don't run arm64 workflows on pull request yet.

I do share your concern around this conditional hack though.

**Comment by serathius** [comment]
tldr; we need to update the release process to incorporate those workflow failures as a signal.

I think we are stomping into new ground. Periodic workflows themselves are not enough to guarantee that next etcd release will work on ARM. They will run, fail but not result in any action if they fail. We need to make them somehow actionable.

To paint the picture. Example of what I mean is process that K8s has. First a dashboard that gives clear view of all periodic tests for all supported versions. Dashboard for v1.24 testing I own https://k8s-testgrid.appspot.com/sig-instrumentation-tests#kind-json-logging-1.24
If there is a repeated test failure, an owner of such test scenario will be notified by email. Independently the K8s release team is monitoring same test failures on their dashboard https://k8s-testgrid.appspot.com/sig-release-1.27-blocking#gce-cos-k8sstable1-default. If a particular test failure blocks a release, they will ping the test owner to fix it.
Before making a release they check that all runs are green.

I think to make the periodic workflows failures be a signal for release we need to adapt to a process similar to this. Of course we will not try to rollout TestGrid ourselves, but as etcd joins Kubernetes it would be nice to have some integration. We already have a release process and release owners defined, we just need to make results from those periodics a clear signal for our releases.

**Comment by jmhbnz** [comment]
> I think to make the periodic workflows failures be a signal for release we need to adapt to a process similar to this. Of course we will not try to rollout TestGrid ourselves, but as etcd joins Kubernetes it would be nice to have some integration. We already have a release process and release owners defined, we just need to make results from those periodics a clear signal for our releases.

Completely agree.  Additionally I am intending to scale up our runners in future to have multiple on each box (we have tons of compute idle on each box) so that we can actually run some of these workflows on each pr also.

File: .github/workflows/tests-arm64-nightly.yaml (hunk 1 above)
Comment (serathius) [reply .github/workflows/tests-arm64-nightly.yaml:21]: Yes, I mean backport the makefile changes to keep a single interface for all branches. It should the forkflows easier to maintain and simplify development of backports. cc @ahrtr

Comment (jmhbnz) [reply .github/workflows/tests-arm64-nightly.yaml:21]: I'm happy to create an issue and work on it if we have consensus from others.  I'm not sure I would want it to be a blocking issue for this pr though. Keen to see what overall consensus is 🙂

Comment (ahrtr) [reply .github/workflows/tests-arm64-nightly.yaml:21]: > Yes, I mean backport the makefile changes to keep a single interface for all branches.


Makes sense to me.

File: .github/workflows/e2e-arm64-template.yaml [hunk 4]
```diff
@@ -1,8 +1,14 @@
 ---
-name: E2E-arm64
+name: Reusable Arm64 E2E Workflow
```
Comment (ahrtr) [inline .github/workflows/e2e-arm64-template.yaml:2]: Probably I am missing something, the template is only used by e2e-arm64-nightly for now. In other words, the template has only one user, then why do you need to create a template? If the template can be reused by both ARM e2e test and NON-ARM e2e test, then it makes sense.

Comment (jmhbnz) [reply .github/workflows/e2e-arm64-template.yaml:2]: We use a re-usable workflow so we can run the same template for `main` and `release-3.5` with different parameters. Similar to how robustness workflows run across different branches using the same workflow template.

Comment (ahrtr) [reply .github/workflows/e2e-arm64-template.yaml:2]: thx for the clarification. It seems it would be better if there is a document for all the workflow tests.

File: .github/workflows/tests-arm64-nightly.yaml (hunk 1 above)
Comment (jmhbnz) [reply .github/workflows/tests-arm64-nightly.yaml:21]: Created https://github.com/etcd-io/etcd/issues/16176 under our overall ci improvements plan to track.  I am keen to work on this once we have tier 1 arm64 announced 🙏🏻

**Comment by jmhbnz** [comment]
Hey @serathius, @ahrtr - Can you please take another look at this? I'm keen to see how these tests will behave against 3.5.

**Review by ahrtr — APPROVED** [review]
LGTM

Thanks @jmhbnz 

I think we can discuss the adapting the similar K8s process in a separate session. Leave to @serathius to approve and merge this PR.

**Review by serathius — APPROVED** [review]
